### ISSUE TITLE
FR: Change Background color/shade

### ISSUE THREAD

**Issue by gerroon** [koreader/koreader#3871]
Hi

It would be so nice if one could change the background color of the app, especially in the content area. It is set to pure white and sometimes not all content fits nicely to the screen so there is some white leaking through edges which gets a bit distracting. This is a bigger issue when used on Lcd devices.

I recommend having at least 3 basic shades like white, gray and black. The menus can stay as is, this is purely about the content area where the book pages are rendered over. I am not talking about background color of the books.

**Comment by Frenzie** [comment]
If/when changing this it'd make no sense to limit it artificially, except in preset menu options.

Since white is hard-coded in so many places for now it's basically impossible except by redefining white. And more changes would be needed to affect the documents themselves. A stretch goal for now, not a quickie. :-)

![screenshot_2018-04-13_19-39-21](https://user-images.githubusercontent.com/202757/38749632-960b58d0-3f52-11e8-94a9-139f28b66f0b.png)

**Comment by gerroon** [comment]
Thanks for considering.

**Comment by poire-z** [comment]
Difficult indeed to have the whole thing consistent.

But in case you want it for epubs/crengine docs, you can set one in `data/cr3.ini`, by replacing:
`background.color.default=0x00FFFFFF` with something like background.color.default=0x00BBFFCC

On Android, I remember i had it working with this in patch.lua (may be it doesn't work anymore):
```lua
-- Set CREngine documents background color
local called_once = false
local fba = require("ffi/framebuffer_android")
fba.refreshFullImp_orig = fba.refreshFullImp
fba.refreshFullImp = function(self)
    fba.refreshFullImp_orig(self)
    if not called_once then
        called_once = true
        return
    end
    fba.refreshFullImp = fba.refreshFullImp_orig
    local CreDocument = require("document/credocument")
    CreDocument.initOrig = CreDocument.init
    CreDocument.init = function(self)
        CreDocument.initOrig(self)
        self._document:setStringProperty("background.color.default", "#eeeeee")
    end
end
```

**Comment by gerroon** [comment]
@poire-z 

I mostly use it to read df files. Will the patch work for pdfs?

**Comment by poire-z** [comment]
No, it won't work. That was easy for epub because crengine internally supports this feature of background color.
I know not much about MuPDF, but I doubt it does.

**Comment by Frenzie** [comment]
StackOverflow result: https://stackoverflow.com/a/15007540/2470572

Alternatively (skimming the headers) maybe something with `fz_shade`?

**Comment by pazos** [comment]
For android it is possible to draw an almost transparent overlay and set some color themes, like this [PDF Expert example](https://helpspot.readdle.com/pdfexpert6/index.php?pg=kb.page&id=1112)

Of course it is not the same as a background color, because it will affect everything rendered on screen.

**Comment by ildar** [comment]
What about PDF reflow mode? Should be relatively easy

**Comment by KenMaltby** [comment]
CoolReader had it so you could have a textured look background, most often a parchment look.

**Comment by poire-z** [comment]
Just a note about crengine: we can set some background color via cr3.ini, but for a background image, it would need some function in cre.cpp, doing something like:
```c
    LVStreamRef stream = LVOpenFileStream("/tmp/some_texture.png", LVOM_READ);
    if ( !stream.isNull() ) {
        LVImageSourceRef img = LVCreateStreamCopyImageSource(stream);
        if ( !img.isNull() ) {
            doc->text_view->setBackgroundImage(img, true);
        }
    }
```
The menu and all the frontend UI elements would still have a white background - and when only reading, the footer would still have a white background.

**Comment by pbienst** [comment]
What's the current hack to set the background colour on Android for epubs? Neither data/cr3.ini nor patch.lua seems to do anything...

**Comment by poire-z** [comment]
Dunno, and just because I don't want to test and check what works or not, I'll add (in a few days) support for 2 settings one can add manually in his settings.reader.lua:
```lua
    ["cre_background_color"] = "0xccffcc",
    ["cre_background_image"] = "/mnt/onboard/.adds/bg_paper.jpg",
```

**Comment by pbienst** [comment]
Thanks!

**Comment by mendelct** [comment]
The catch with changing the ["cre_background_color"] is that when going into night mode you invert whatever the background is so if you do, for example, a tan background it will invert to blue. To get around that I patch the CreDocument:drawCurrentView function (in credocument.lua in frontend/document) with

`

local background_color = self._document:getStringProperty("background.color.default")

    if  Screen.night_mode then

	if  background_color ~= "#ffffff" then

          self._document:setStringProperty("background.color.default", "#ffffff")

	end

    else

	if background_color ~= "#d2b48c" then

          self._document:setStringProperty("background.color.default", "#d2b48c")

	end

    end
`


I hardcoded the values I like. It should be fairly simple to make a UI to allow a colorpicker. 

When night mode is selected for files being viewed by crengine the first page will be an invert of the previous color (I assume b/c it is just inverted not redrawn), but the next page will be the invert of the background color set in the code. If night mode is deselected the first page will be the invert of night mode but the next page will be the color set in the code. 
I would like the ability to set the font color as well but that is a minor detail.

for pdfs I patch the run_page function (in mupdf.lua in base/ffi) . The following is inserted after clear_pixmap:

`
if not Screen.night_mode then

	r = 210
	g = 180
	b = 140
	M.fz_tint_pixmap(context(), pixmap, r, g, b);
end
`

You need to require the screen `local Screen = require("device").screen`

You also need to declare the function in mupdf_h.lua
`void fz_tint_pixmap(fz_context *, fz_pixmap *, int r, int g, int b);`

For pdfs, however, when changing to night mode any page that was already viewed will show up as inverted, even after a restart. I'm assuming that the pages are cached and are not redrawn/rerun. Any suggestions how to require all pages to be redrawn/rerun after a night mode change would be appreciated.

**Comment by Frenzie** [comment]
Any property that should result in a redraw has be made part of the hash. Mainly but not only:

https://github.com/koreader/koreader/blob/2e947e6ffe5dadbb7f3b1ef7430bf91b4f51d9fe/frontend/document/document.lua#L314-L318

**Comment by Thertzlor** [comment]
While the "cre_background_color" option works well for the actual book contents it would still be nice to be able to change the color of menu elements.
Especially the reader footer remaining white when the background for the rest of the screen has changed is a bit off-putting.

It seems relevant to bring this up now since a few releases ago I was still able to override the footer background and a few other colors with the patch.lua file by modifying the package.path and forcing the app to load a modified readerfooter.lua file  from the very start but it seems like this approach no longer works with the new "patches" folder.

Is there perhaps any other way to tweak interface colors with a userscript?

**Comment by poire-z** [comment]
Dunno what was/is your patch, but:
`koreader/patches/2-color-white-not-white.lua`
```lua
local BB = require("ffi/blitbuffer")
BB.COLOR_WHITE = BB.Color8(0xCC)
```
would give:
![image](https://user-images.githubusercontent.com/24273478/202796475-1b10ab26-43bd-4ab0-8b66-cd4f89b02553.png)

https://github.com/koreader/koreader/pull/3285#issuecomment-336248237

Also, our UI painting (font, filling) code only supports 8bit colors, so the UI bg and fg color only supports grays. https://github.com/koreader/koreader/pull/5854#issuecomment-585216764

**Comment by Thertzlor** [comment]
Thanks for the script; It works fine but it's a much more general effect than the replacement I had before.
And yes, I am aware of the 8bit restriction, I did manage to circumvent it by messing with the "getColor8()" calls in various files (the readerfooter.lua wasn't the only thing I replaced). But with less saturated colors choosing a grey tone that matches the lightness of the background looks good enough, so I think I'll just stick with that...

**Comment by poire-z** [comment]
That was just a suggestion to show that you can achieve _something_ with a user patch.
You're still not showing what was your previous patch. But that's good, as I then won't be tempted to make it work for you :)
Anyway, look at the existing user patches, https://github.com/koreader/koreader/wiki/User-patches.
It might be simpler than what you needed to do before.
Ie, if you would need to use a hacked getColor8 only in some specific existing methods, you could use tricks like:
```lua
local orig_getColor8 = bb.getColor8
local my_getColor8 = function(...) ... end
local orig_blah = ReaderFooter.blah
ReaderFooter.blah = function()
  bb.getColor8 = my_getColor8
  orig_blah(self)
  bb.getColor8 = orig_getColor8
end
```

**Comment by shuvashish76** [comment]
@Euphoriyy 's user patches for KOReader's UI font and background-color with color picker option : https://github.com/Euphoriyy/KOReader.patches?tab=readme-ov-file#-2-ui-font-colorlua

Please add "user patch available" label.

Edit: Now available as a single plugin: https://github.com/Euphoriyy/appearance.koplugin
